### PR TITLE
Dt docs for 2.3 part 2

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Addresses these issues from #25469:

- review doxygen docstrings and correct errors for instance APIs
- review doxygen docstrings and correct errors for hw specific APIs
- address outstanding comments from #25009
- review user guides and correct errors

There will be a part 3, but that will be the last one.

### CODE REVIEW

File: doc/guides/dts/api-usage.rst [hunk 1]
```diff
@@ -8,37 +8,60 @@ from C source files. It assumes you're familiar with the concepts in
...
+
+By instance number
+   This is done primarily by device drivers, as instance numbers are a way to
+   refer to individual nodes based on a matching compatible. Get these with
+   :c:func:`DT_INST()`, but be careful doing so. See below.
+
+By chosen node
+   Use :c:func:`DT_CHOSEN()` to get a node identifier for ``/chosen`` node
+   properties.
+
```
Comment (galak) [inline doc/guides/dts/api-usage.rst:62]: should we add parent and child in here?

Comment (mbolivar-nordic) [reply doc/guides/dts/api-usage.rst:62]: I can if you think we should, but I was hoping to keep the list as small as possible for an introductory guide. There are other cases that return node identifiers -- devicetree/spi.h has at least one. Figured we'd leave those as "advanced" for now.

File: doc/guides/dts/design.rst [hunk 2]
```diff
@@ -78,9 +78,8 @@ Example remaining work
   dtschema does not yet meet Zephyr's needs, we should try to follow what it is
   capable of representing in Zephyr's own bindings.
 
-- Due to namespace collisions and inflexibility in the bindings language,
-  Zephyr cannot support the full set of possible node paths and bindings
-  supported by Linux.
+- Due to inflexibility in the bindings language, Zephyr cannot support the full
+  set of bindings supported by Linux.
```
Comment (galak) [inline doc/guides/dts/design.rst:82]: what is this trying to cover?  Not issue with the intent, just not exactly sure what case we have issues w/from a binding point of view.

File: doc/guides/dts/howtos.rst [hunk 3]
```diff
@@ -113,7 +113,7 @@ that the node is has ``status = "okay"`` and has a matching binding, like this:
    #if DT_NODE_HAS_STATUS(MY_SERIAL, okay)
    struct device *uart_dev = device_get_binding(DT_LABEL(MY_SERIAL));
    #else
-   #error "Node is disabled, has no matching binding, or initialization failed"
+   #error "Node is disabled, has no matching binding, or initialization failed"X
```
Comment (galak) [inline doc/guides/dts/howtos.rst:116]: is trailing `X` here correct?

File: include/devicetree/adc.h [hunk 4]
```diff
@@ -23,10 +23,10 @@ extern "C" {
  */
 
 /**
- * @brief Get IO channels controller "name" (label property) at an index
+ * @brief Get a label property from an io-channels property at an index
```
Comment (galak) [inline include/devicetree/adc.h:26]: `Get a label property from the node referenced by io-channels property at an index`?

File: include/devicetree/clocks.h [hunk 5]
```diff
@@ -23,10 +23,11 @@ extern "C" {
  */
 
 /**
- * @brief Get clock controller "name" (label property) at an index
+ * @brief Get a label property from a clocks property at an index
```
Comment (galak) [inline include/devicetree/clocks.h:26]: same comment as for adc.h

File: include/devicetree/dma.h [hunk 6]
```diff
@@ -17,16 +17,16 @@ extern "C" {
...
 /**
- * @defgroup devicetree-dmas Devicetree dma API
+ * @defgroup devicetree-dmas Devicetree DMA API
  * @ingroup devicetree
  * @{
  */
 
 /**
- * @brief Get dma controller "name" (label property) at an index
+ * @brief Get a label property from a dmas property at an index
```
Comment (galak) [inline include/devicetree/dma.h:26]: same comment as adc.h

**Review by galak — COMMENTED** [review]
Looks mostly good, a few comments here are there.

File: doc/guides/dts/design.rst (hunk 2 above)
Comment (mbolivar-nordic) [reply doc/guides/dts/design.rst:82]: Linux is more general than Zephyr when handling specifier spaces and nexus nodes at least.

For example, our handling of `#foo-cells` is done by just stripping the trailing `-s` from `foos`:

https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/dts/edtlib.py#L1417

While convenient and though it gets the job done for now, it's not sufficiently expressive for the `mboxes` / `#mbox-cells` binding for example:

https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/mailbox/mailbox.txt#L9

Our handling of the "compound" type (i.e. generate nothing) is another example that holds us back from expressing nexus nodes with the same level of generality as Linux, IIUC.

**Comment by mbolivar-nordic** [comment]
Addressed, thanks

**Review by galak — APPROVED** [review]

File: doc/guides/dts/api-usage.rst [hunk 7]
```diff
@@ -8,37 +8,64 @@ from C source files. It assumes you're familiar with the concepts in
...
+A note for Linux developers
+***************************
+
+Linux developers familiar with devicetree should be warned that the API
+described here differs significantly from how devicetree is used on Linux.
+
+Instead of generating a C header with all the devicetree data which is then
+abstracted behind a macro API, the Linux kernel would instead read the
+devicetree data structure in its binary form. The binary representation is
+parsed it at runtime, for example to load and initialize device drivers.
```
Comment (henrikbrixandersen) [inline doc/guides/dts/api-usage.rst:20]: s/parsed it at/parsed at/

**Review by henrikbrixandersen — APPROVED** [review]
This is a large update to review, but from my one-pass read-through, it looks very good (apart from one very minor nit).

**Comment by mbolivar-nordic** [comment]
> This is a large update to review, but from my one-pass read-through, it looks very good (apart from one very minor nit).

Thank you, now addressed.

**Review by carlescufi — APPROVED** [review]
Excellent work

**Review by pabigot — APPROVED** [review]
Looks great; thanks.

**Pull request merged**